components/brave_shields/browser/brave_shields_web_contents_observer.cc

```diff
@@ -255,6 +261,14 @@ void BraveShieldsWebContentsObserver::DispatchBlockedEvent(
             prefs->GetUint64(kFingerprintingBlocked) + 1);
       }
     }
+
+#if BUILDFLAG(ENABLE_BRAVE_PERF_PREDICTOR)
```

Comment:
How about calling this directly from `DispatchBlockedEvent()` instead of calling via this shields tab helper?
I think both TabHelper have its own responsibility. WDYT?

Reply:
short of having a separate dispatcher like those for `WebContentsObserver` or `PageLoadMetricsObserver` to actually decouple components it doesn't seem to make a big difference. Moving it to the util `DispatchBlockedEvent` would require a repeat of the logic to get hold of `web_contents` as well

Reply:
What I'm concerning here is modifying `BraveShieldsWebContentsObserver` logic should not affect `PerfPredictorTabHelper` because both do its own job. I think there is no reason to couple both TabHelper.
I think we can change `OnBlockedSubresource` method signature and fetching `web_contents` from there.

Reply:
+1 for decoulping things

Reply:
all up for decoupling, but `brave_shields` component doesn't notify anything outside the component as far as I can tell, so not following the part with making `OnBlockedSubresource` static? How would you obtain the correct instance of `BandwidthPredictor` (which accumulates state over a page load) from a static method there? What calls that static method?

Reply:
I think this will work.
In `brave_shields_util.cc`
```
void DispatchBlockedEvent(const GURL& request_url,
                          int render_frame_id,
                          int render_process_id,
                          int frame_tree_node_id,
                          const std::string& block_type) {
  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
  BraveShieldsWebContentsObserver::DispatchBlockedEvent(
      block_type, request_url.spec(),
      render_process_id, render_frame_id, frame_tree_node_id);

  brave_perf_predictor::PerfPredictorTabHelper::DispatchBlockedEvent(
      request_url.spec(), render_process_id,
      render_frame_id, frame_tree_node_id
  );
}
```

and In `perf_predictor_tab_helper.cc`
```
namespace {
WebContents* GetWebContents(
    int render_process_id,
    int render_frame_id,
    int frame_tree_node_id) {
  WebContents* web_contents =
      WebContents::FromFrameTreeNodeId(frame_tree_node_id);
  if (!web_contents) {
    RenderFrameHost* rfh =
        RenderFrameHost::FromID(render_process_id, render_frame_id);
    if (!rfh) {
      return nullptr;
    }
    web_contents =
        WebContents::FromRenderFrameHost(rfh);
  }
  return web_contents;
}

}  // namespace

// static
void PerfPredictorTabHelper::DispatchBlockedEvent(std::string subresource,
                                                  int render_process_id,
                                                  int render_frame_id,
                                                  int frame_tree_node_id) {
  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);

  content::WebContents* web_contents = GetWebContents(
      render_process_id, render_frame_id, frame_tree_node_id);
  PerfPredictorTabHelper* blocking_observer =
      brave_perf_predictor::PerfPredictorTabHelper::FromWebContents(
          web_contents);
  if (blocking_observer) {
    blocking_observer->OnBlockedSubresource(subresource);
  }
 }
```